pkg/tasks/cleanup_trashcans_service.go

```diff
@@ -0,0 +1,88 @@
+package tasks
+
+import (
+	"context"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/ottogroup/penelope/pkg/http/impersonate"
+	"github.com/ottogroup/penelope/pkg/repository"
+	"github.com/ottogroup/penelope/pkg/secret"
+	"github.com/ottogroup/penelope/pkg/service/gcs"
+	"go.opencensus.io/trace"
+	"time"
+)
+
+func newCleanupTrashcansService(ctxIn context.Context, tokenSourceProvider impersonate.TargetPrincipalForProjectProvider, credentialsProvider secret.SecretProvider) (*cleanupTrashcansService, error) {
+	ctx, span := trace.StartSpan(ctxIn, "newPrepareBackupJobsService")
+	defer span.End()
+
+	backupRepository, err := repository.NewBackupRepository(ctx, credentialsProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cleanupTrashcansService{tokenSourceProvider: tokenSourceProvider, backupRepository: backupRepository}, nil
+}
+
+type cleanupTrashcansService struct {
+	tokenSourceProvider impersonate.TargetPrincipalForProjectProvider
+	backupRepository    repository.BackupRepository
+}
+
+func (s *cleanupTrashcansService) Run(ctxIn context.Context) {
+	ctx, span := trace.StartSpan(ctxIn, "(*cleanupTrashcansService).Run")
+	defer span.End()
+
+	backups, err := s.backupRepository.GetBackupsByCleanupTrashcanStatus(ctx, repository.ScheduledTrashcanCleanupStatus)
+	if err != nil {
+		glog.Errorf("could not get list of backups scheduled to cleanup trashcan: %s", err)
+	}
+
+	for _, backup := range backups {
+		gcsClient, err := gcs.NewCloudStorageClient(ctx, s.tokenSourceProvider, backup.TargetProject)
+		if err != nil {
+			glog.Errorf("could not create new CloudStorageClient: %s", err)
+			return
+		}
+		defer gcsClient.Close(ctx)
+
+		err = s.backupRepository.MarkTrashcanCleanup(ctx, backup.ID, repository.TrashcanCleanup{
+			Status:                   repository.InProgressCleanupTrashcanCleanupStatus,
+			StartInProgressTimestamp: time.Now(),
+		})
+		if err != nil {
+			return
+		}
+
+		err = gcsClient.DeleteObjectWithPrefix(ctx, backup.Sink, backup.GetTrashcanPath())
```

Comment:
I have the urge to protect DeleteObjectWithPrefix against empty result of GetTrashcanPath to not delete the complete backup

Reply:
Setting trashcan status to error if `backup.GetTrashcanPath()` returns empty string